### PR TITLE
Renamed SafeCore to Anchor

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR does following :-
1. Renames SafeCore contract name to Anchor.
2. Updates the references of SafeCore in unit test cases.
3. Changes the method name commitStateRoot to anchorStateRoot in Anchor contract. 
4. Changes the method name setCoCoreAddress to setCoAnchorAddress in Anchor contract.

Fixes: #537

### CODE REVIEW

File: contracts/gateway/Anchor.sol [hunk 1]
```diff
@@ -30,15 +29,18 @@ import "../lib/RLP.sol";
 import "../lib/SafeMath.sol";
 
 /**
- * @title SafeCore contract which implements StateRootInterface.
+ * @title Anchor contract which implements StateRootInterface.
  *
- * @notice SafeCore stores another chain's state roots. It stores the address of
- *         the co-core, which will be the safe core on the other chain. State
+ * @notice Anchor stores another chain's state roots. It stores the address of
+ *         the co-core, which will be the anchor on the other chain. State
```
Comment (schemar) [inline contracts/gateway/Anchor.sol]:
```suggestion
 *         the co-anchor, which will be the anchor on the other chain. State
```

File: test/gateway/anchor/get_latest_state_root_block_height.js [hunk 2]
```diff
@@ -41,8 +41,8 @@ contract('SafeCore.getLatestStateRootBlockHeight()', function (accounts) {
     blockHeight = new BN(5);
     stateRoot = web3.utils.sha3("dummy_state_root");
     membersManager = await MockMembersManager.new(owner, worker);
-
-    safeCore = await SafeCore.new(
+
```
Comment (schemar) [inline test/gateway/anchor/get_latest_state_root_block_height.js]:
```suggestion

```

File: test/gateway/anchor/set_co_core_address.js [hunk 3]
```diff
@@ -46,8 +46,8 @@ contract('SafeCore.setCoCoreAddress()', function (accounts) {
     stateRoot = web3.utils.sha3("dummy_state_root");
     membersManager = await MockMembersManager.new(owner, worker);
     coCoreAddress = accounts[6];
-
-    safeCore = await SafeCore.new(
+
```
Comment (schemar) [inline test/gateway/anchor/set_co_core_address.js]:
```suggestion

```

File: test/gateway/anchor/set_co_core_address.js [hunk 4]
```diff
@@ -61,7 +61,7 @@ contract('SafeCore.setCoCoreAddress()', function (accounts) {
     coCoreAddress = NullAddress;
 
     await Utils.expectRevert(
-      safeCore.setCoCoreAddress(coCoreAddress, { from: owner }),
+        anchor.setCoCoreAddress(coCoreAddress, { from: owner }),
```
Comment (schemar) [inline test/gateway/anchor/set_co_core_address.js]: In many places:
```suggestion
      anchor.setCoCoreAddress(coCoreAddress, { from: owner }),
```

File: contracts/gateway/Anchor.sol [hunk 5]
```diff
@@ -30,15 +29,18 @@ import "../lib/RLP.sol";
...
  *         roots are exchanged bidirectionally between the core and the co-core
```
Comment (schemar) [inline contracts/gateway/Anchor.sol]:
```suggestion
 *         roots are exchanged bidirectionally between the anchor and the co-anchor
```

File: contracts/gateway/Anchor.sol [hunk 6]
```diff
@@ -156,18 +158,18 @@ contract SafeCore is StateRootInterface, Organized {
     }
 
     /**
-     *  @notice External function commitStateRoot.
+     *  @notice External function anchorStateRoot.
      *
-     *  @dev commitStateRoot Called from game process.
-     *       Commit new state root for a block height.
+     *  @dev anchorStateRoot Called from game process.
+     *       Anchor new state root for a block height.
```
Comment (schemar) [inline contracts/gateway/Anchor.sol]: Please clean up these doc comments. they don't really fit. The `@notice` doesn't help and the `@dev` is also somewhat redundant. Maybe remove the `@dev` and make the `@notice`:

> Anchor a new state root for a given block height.

File: contracts/gateway/Anchor.sol [hunk 7]
```diff
@@ -184,7 +184,7 @@ contract SafeCore is StateRootInterface, Organized {
         // Input block height should be valid
```
Comment (sarvesh-ost) [inline contracts/gateway/Anchor.sol]: Please add `.` in the comments.

File: contracts/test/core/MockAnchor.sol [hunk 8]
```diff
@@ -1,24 +1,23 @@
...
+contract MockAnchor is Anchor {
 
     /*  Public functions */
 
     /**
-     * @notice Contract constructor
+     * @notice Contract constructor.
      *
      *  @param _remoteChainId _remoteChainId is the chain id of the chain that
      *                        this core tracks.
```
Comment (sarvesh-ost) [inline contracts/test/core/MockAnchor.sol]:
```suggestion
      *                        this anchor tracks.
```

File: contracts/test/core/MockAnchor.sol [hunk 9]
```diff
@@ -27,7 +26,7 @@ contract MockSafeCore is SafeCore {
         bytes32 _stateRoot,
         IsMemberInterface _membersManager
     )
-        SafeCore(
+        Anchor(
```
Comment (sarvesh-ost) [inline contracts/test/core/MockAnchor.sol]: Please update comment on L42 
`note the Core contract` -> `note the anchor contract`

File: contracts/test/gateway/MockGatewayBase.sol [hunk 10]
```diff
@@ -42,7 +42,7 @@ contract MockGatewayBase is GatewayBase {
      *
      *  @dev proveGateway can be called by anyone to verify merkle proof of
      *       gateway/co-gateway contract address. Trust factor is brought by
-     *       stateRoots mapping. stateRoot is committed in commitStateRoot
+     *       stateRoots mapping. stateRoot is committed in anchorStateRoot
```
Comment (sarvesh-ost) [inline contracts/test/gateway/MockGatewayBase.sol]:
```suggestion
     *       stateRoots mapping. stateRoot is anchored in anchorStateRoot
```

File: test/gateway/anchor/get_latest_state_root_block_height.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        let latestBlockHeight = await anchor.getLatestStateRootBlockHeight.call();
+        assert.strictEqual(
+            blockHeight.eq(latestBlockHeight),
+            true,
+            `Latest block height from the contract must be ${blockHeight}.`,
+        );
+        
+    });
+    
+    it('should return the latest committed state root block height', async () => {
```
Comment (sarvesh-ost) [inline test/gateway/anchor/get_latest_state_root_block_height.js]:
```suggestion
     it('should return the latest anchored state root block height', async () => {
```

File: test/gateway/anchor/get_state_root.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        let latestStateRoot = await anchor.getStateRoot.call(blockHeight);
+        assert.strictEqual(
+            latestStateRoot,
+            zeroBytes,
+            `Latest state root from the contract must be ${zeroBytes}.`,
+        );
+        
+    });
+    
+    it('should return the latest committed state root', async () => {
```
Comment (sarvesh-ost) [inline test/gateway/anchor/get_state_root.js]:
```suggestion
     it('should return the latest anchored state root', async () => {
```

**Comment by schemar** [comment]
> Also fix terminology in below contracts:
> * StateRootInterface
> * GatewayLib
> * GatewayBase
> * MockAnchor L18, L42

I don't think the gateway code should be aware whether it's an anchor or a core. The gateways should only care about the state root interface. E.g. "state root provider". Or am I missing something?

**Comment by sarvesh-ost** [comment]
> > Also fix terminology in below contracts:
> > 
> > * StateRootInterface
> > * GatewayLib
> > * GatewayBase
> > * MockAnchor L18, L42
> 
> I don't think the gateway code should be aware whether it's an anchor or a core. The gateways should only care about the state root interface. E.g. "state root provider". Or am I missing something?
> 
> @gulshanvasnani @sarvesh-ost

What should be the name of the variable which represents contract instance implementing `StateRootInterface`? 
Currently, it's `core` which seems incorrect, as `Anchor` contract is providing state roots. 
Do we need third terminology other than `core` and `Anchor`? 🤔

**Comment by schemar** [comment]
> > > Also fix terminology in below contracts:
> > > 
> > > * StateRootInterface
> > > * GatewayLib
> > > * GatewayBase
> > > * MockAnchor L18, L42
> > 
> > 
> > I don't think the gateway code should be aware whether it's an anchor or a core. The gateways should only care about the state root interface. E.g. "state root provider". Or am I missing something?
> 
> What should be the name of the variable which represents contract instance implementing `StateRootInterface`?
> Currently, it's `core` which seems incorrect, as `Anchor` contract is providing state roots.
> Do we need third terminology other than `core` and `Anchor`? 🤔

Yes, we do. As it can be both. It can be an anchor as well as a core. In fact, it can be anything that implements the state root interface. It's just that in the past only *different kinds of cores* implemented that interface. Now that the safe core is called anchor, we need a more general name for the variable of that interface type.

**Comment by gulshanvasnani** [comment]
> > > > Also fix terminology in below contracts:
> > > > 
> > > > * StateRootInterface
> > > > * GatewayLib
> > > > * GatewayBase
> > > > * MockAnchor L18, L42
> > > 
> > > 
> > > I don't think the gateway code should be aware whether it's an anchor or a core. The gateways should only care about the state root interface. E.g. "state root provider". Or am I missing something?
> > 
> > 
> > What should be the name of the variable which represents contract instance implementing `StateRootInterface`?
> > Currently, it's `core` which seems incorrect, as `Anchor` contract is providing state roots.
> > Do we need third terminology other than `core` and `Anchor`? 🤔
> 
> Yes, we do. As it can be both. It can be an anchor as well as a core. In fact, it can be anything that implements the state root interface. It's just that in the past only _different kinds of cores_ implemented that interface. Now that the safe core is called anchor, we need a more general name for the variable of that interface type.

Name can be stateRoots

File: contracts/StateRootInterface.sol [hunk 13]
```diff
@@ -24,9 +24,9 @@ pragma solidity ^0.5.0;
 interface StateRootInterface {
 
     /**
-     * @notice Gets the block number of latest committed state root.
+     * @notice Gets the block number of latest anchored state root.
```
Comment (schemar) [inline contracts/StateRootInterface.sol]: `committed` is actually more appropriate here. The state root interface is *not* specific to the anchor.

File: contracts/StateRootInterface.sol [hunk 14]
```diff
@@ -24,9 +24,9 @@ pragma solidity ^0.5.0;
...
      *
-     * @return uint256 Block height of the latest committed state root.
+     * @return uint256 Block height of the latest anchored state root.
```
Comment (schemar) [inline contracts/StateRootInterface.sol]:
```suggestion
     * @return height_ Block height of the latest anchored state root.
```

File: contracts/gateway/EIP20CoGateway.sol [hunk 15]
```diff
@@ -214,14 +214,14 @@ contract EIP20CoGateway is GatewayBase {
...
-        StateRootInterface _core,
+        StateRootInterface _anchor,
         uint256 _bounty,
         IsMemberInterface _membersManager,
         address _gateway,
         address payable _burner
     )
         GatewayBase(
-            _core,
+            _anchor,
```
Comment (schemar) [inline contracts/gateway/EIP20CoGateway.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/gateway/EIP20Gateway.sol [hunk 16]
```diff
@@ -240,13 +240,13 @@ contract EIP20Gateway is GatewayBase {
...
         EIP20Interface _baseToken,
-        StateRootInterface _core,
+        StateRootInterface _anchor,
         uint256 _bounty,
         IsMemberInterface _membersManager,
         address _burner
     )
         GatewayBase(
-            _core,
+            _anchor,
```
Comment (schemar) [inline contracts/gateway/EIP20Gateway.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/gateway/GatewayBase.sol [hunk 17]
```diff
@@ -74,8 +74,8 @@ contract GatewayBase is Organized {
      */
     MessageBus.MessageBox messageBox;
 
-    /** Address of core contract. */
-    StateRootInterface public core;
+    /** Address of anchor contract. */
+    StateRootInterface public anchor;
```
Comment (schemar) [inline contracts/gateway/GatewayBase.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/lib/GatewayLib.sol [hunk 18]
```diff
@@ -126,7 +126,7 @@ library GatewayLib {
         );
 
         /**
-         * Verify the remote OpenST contract against the committed state
+         * Verify the remote OpenST contract against the anchored state
```
Comment (schemar) [inline contracts/lib/GatewayLib.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/test/TestEIP20CoGateway.sol [hunk 19]
```diff
@@ -57,7 +57,7 @@ contract TestEIP20CoGateway is EIP20CoGateway {
         EIP20CoGateway(
             _valueToken,
             _utilityToken,
-            _core,
+            _anchor,
```
Comment (schemar) [inline contracts/test/TestEIP20CoGateway.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/test/TestEIP20Gateway.sol [hunk 20]
```diff
@@ -45,15 +45,15 @@ contract TestEIP20Gateway is EIP20Gateway {
...
+        StateRootInterface _anchor,
         uint256 _bounty,
         IsMemberInterface _membersManager,
         address payable _burner
     )
         EIP20Gateway(
             _token,
             _baseToken,
-            _core,
+            _anchor,
```
Comment (schemar) [inline contracts/test/TestEIP20Gateway.sol]: Also in this file it should not be called anchor, but something more general.

File: contracts/test/gateway/MockGatewayBase.sol [hunk 21]
```diff
@@ -17,19 +17,19 @@ contract MockGatewayBase is GatewayBase {
     /**
      * @notice This is used for testing.
      *
-     * @param _core Core contract address.
+     * @param _anchor Anchor contract address.
```
Comment (schemar) [inline contracts/test/gateway/MockGatewayBase.sol]: Also in this file it should not be called anchor, but something more general.

File: test/gateway/anchor/anchor_state_root.js [hunk 22]
```diff
@@ -28,171 +28,171 @@ const EventDecoder = require('../../test_lib/event_decoder.js');
...
+    
+    anchor = await Anchor.new(
       remoteChainId,
       blockHeight,
       stateRoot,
       maxNumberOfStateRoots,
       membersManager.address,
     );
-
+
```
Comment (schemar) [inline test/gateway/anchor/anchor_state_root.js]: Never end lines with extra spaces

File: test/gateway/anchor/anchor_state_root.js [hunk 23]
```diff
@@ -28,171 +28,171 @@ const EventDecoder = require('../../test_lib/event_decoder.js');
...
-          { from: worker },
-        ),
-        'Given block height is lower or equal to highest committed state root block height.',
-      );
-
-    });
-
-  it('should fail when block height is equal to the latest committed ' +
+    
+    blockHeight = blockHeight.subn(1);
```
Comment (schemar) [inline test/gateway/anchor/anchor_state_root.js]: indentation was correct before, is now wrong.

File: test/gateway/eip20_cogateway/constructor.js [hunk 24]
```diff
@@ -30,15 +30,15 @@ const NullAddress = "0x0000000000000000000000000000000000000000";
 contract('EIP20CoGateway.constructor() ', function (accounts) {
 
 
-    let valueToken, utilityToken, bountyAmount, coreAddress, owner, worker,
+    let valueToken, utilityToken, bountyAmount, anchorAddress, owner, worker,
```
Comment (schemar) [inline test/gateway/eip20_cogateway/constructor.js]: Also in this file it should not be called anchor, but something more general.

File: test/gateway/eip20_cogateway/progress_mint.js [hunk 25]
```diff
@@ -57,7 +57,7 @@ async function _setup(accounts) {
   
   valueToken = accounts[0];
   burner = accounts[10];
-  mockSafeCore = accounts[11];
+  mockAnchor = accounts[11];
```
Comment (schemar) [inline test/gateway/eip20_cogateway/progress_mint.js]: This is not a mock anchor. Use a different name.

File: test/gateway/eip20_gateway/activate_gateway.js [hunk 26]
```diff
@@ -19,15 +19,15 @@ contract('EIP20Gateway.activateGateway()', function (accounts) {
 
         let mockToken = accounts[0],
             baseToken = accounts[1],
-            coreAddress = accounts[2],
+            anchorAddress = accounts[2],
```
Comment (schemar) [inline test/gateway/eip20_gateway/activate_gateway.js]: This is neither core nor anchor. It acts as a placeholder for the interface.

File: test/gateway/eip20_gateway/constructor.js [hunk 27]
```diff
@@ -29,14 +29,14 @@ const NullAddress = "0x0000000000000000000000000000000000000000";
...
-    let mockToken, baseToken, bountyAmount, coreAddress, membersManager,
+    let mockToken, baseToken, bountyAmount, anchorAddress, membersManager,
         gateway, owner, worker, burner = NullAddress;
 
     beforeEach(async function () {
 
         mockToken = await MockToken.new();
         baseToken = await MockToken.new();
-        coreAddress = accounts[1];
+        anchorAddress = accounts[1];
```
Comment (schemar) [inline test/gateway/eip20_gateway/constructor.js]: again

File: contracts/gateway/GatewayBase.sol [hunk 28]
```diff
@@ -156,7 +159,7 @@ contract GatewayBase is Organized {
     /**
      *  @notice proveGateway can be called by anyone to verify merkle proof of
      *          gateway/co-gateway contract address. Trust factor is brought by
-     *          stateRoots mapping. stateRoot is committed in commitStateRoot
+     *          stateRoots mapping. stateRoot is anchored in anchorStateRoot
```
Comment (schemar) [inline contracts/gateway/GatewayBase.sol]: `mapping. stateRoot is provided by stateRootProvider ...`
It should not be specific to anchor or core.
Most importantly, this comment doesn't make any sense anymore. The Anchor is by design **not** a decentralized mosaic process!

File: contracts/test/gateway/MockGatewayBase.sol [hunk 29]
```diff
@@ -42,7 +43,7 @@ contract MockGatewayBase is GatewayBase {
      *
      *  @dev proveGateway can be called by anyone to verify merkle proof of
      *       gateway/co-gateway contract address. Trust factor is brought by
-     *       stateRoots mapping. stateRoot is committed in commitStateRoot
+     *       stateRoots mapping. stateRoot is anchored in anchorStateRoot
```
Comment (schemar) [inline contracts/test/gateway/MockGatewayBase.sol]: Same as `GatewayBase`

File: test/gateway/anchor/anchor_state_root.js (hunk 23 above)
Comment (schemar) [reply test/gateway/anchor/anchor_state_root.js]: Still not fixed

File: test/gateway/eip20_cogateway/redeem.js [hunk 30]
```diff
@@ -19,497 +19,497 @@
...
 
 const TestEIP20CoGateway = artifacts.require('TestEIP20CoGateway'),
-    MockSafeCore = artifacts.require('MockSafeCore'),
-    MessageBus = artifacts.require('MessageBus'),
-    UtilityToken = artifacts.require('UtilityToken'),
-    EIP20Token = artifacts.require('EIP20Token'),
-    BN = require('bn.js'),
-    MockToken = artifacts.require('MockToken'),
-    Utils = require("../../test_lib/utils");
+  MockAnchor = artifacts.require('MockAnchor'),
```
Comment (schemar) [inline test/gateway/eip20_cogateway/redeem.js]: Not used, can be removed.

File: test/gateway/anchor/anchor_state_root.js (hunk 23 above)
Comment (schemar) [reply test/gateway/anchor/anchor_state_root.js]: Still not fixed

File: contracts/test/gateway/MockGatewayBase.sol (hunk 29 above)
Comment (schemar) [inline contracts/test/gateway/MockGatewayBase.sol]: Same as `GatewayBase`

File: test/gateway/anchor/anchor_state_root.js (hunk 23 above)
Comment (gulshanvasnani) [reply test/gateway/anchor/anchor_state_root.js]: It is auto done by IDE.

Comment (schemar) [reply test/gateway/anchor/anchor_state_root.js]: OK 👍

**Pull request merged**